### PR TITLE
Make work on functors

### LINKED ISSUE #117: Make work on Functors
right now all the tests check they are not funning on functors.

With the new API in https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/116
we can have `f ⟂ ḟ` and `f ⟂ f̄``, which would be fine API-wise.

### PR DESCRIPTION
**Pull request by mzgubic**
Closes #117 

To do:
- [x] merge and tag [FiniteDifferences PR](https://github.com/JuliaDiff/FiniteDifferences.jl/pull/172), add compat. Needed to generate `rand_tangent` for `f` when testing constructors, i.e. `rand_tangent(Foo)`
- [x] implement `test_frule` for functors
- [x] implement `test_scalar` for functors
- [x] remove `_ensure_not_running_on_functor`

### CODE REVIEW

File: src/testers.jl
Comment (mzgubic) [inline src/testers.jl:204]: This is the main change, i.e. from
`(xs...) -> f(xs...; fkwargs...)` to
`(f, xs...) -> f(xs...; fkwargs...)`
making the `f` an argument. The rest of the changes are adding an extra cotangent for `f` via `rand_tangent`, and renaming the variables from `x̄_something` to `something_cotangent` since it also includes `f̄`

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:105]: We should make  `f` also participage in `auto_primal_and_tangents`
and for the functors one would optionally write: `test_frule(f ⊢  ḟ, ...)`

Comment (oxinabox) [reply src/testers.jl:105]: Yes, we do.
KeyedArrays use call overloading for accessing by key rather than by integer index.
And for those we likely want to use another KeyedArray as the cotangent

Comment (mzgubic) [reply src/testers.jl]: Done, and ready for review

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:105]: It seems like most places (except line 116) that we use this, we could be using 
```suggestion
        primals = primal.(primals_and_tangents)
```
and for line 116 we might want to define a helper closure locally
`call_on_copy(f, xs...) = deepcopy(f)(deepcopy(xs)...; deepcopy(fkwargs)...)`
Which we also need for line 132.

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:108]: note that we need to deep copy `func` here also.
 as well as on line 109, 112, and 116
 
 Possibly this also could use `_is_inferable(call_on_copy(primals...))` ?

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:227]: possibly we don't need to mention the func here.
It should be in the stack trace anyway

File: src/testers.jl
Comment (mzgubic) [reply src/testers.jl]: Not sure how this would work? Otherwise it's ready for re-review

**Comment by mzgubic** [comment]
No clue why the Documentation test fails, and I can't reproduce it locally even though the Julia version, Documenter version, and ChainRulesCore/TestUtils versions are all the same.

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:135]: Isn't this just the same as:
```suggestion
            call_on_copy,
```

File: src/testers.jl
Comment (oxinabox) [inline src/testers.jl:219]: should we do the same trick here, only insted of `call_on_copy` it is just `call`?
Which we would use here and on line 195?

File: src/testers.jl
Comment (oxinabox) [reply src/testers.jl]: Ah year it wouldn't quite work

**Comment by oxinabox** [comment]
> No clue why the Documentation test fails, and I can't reproduce it locally even though the Julia version, Documenter version, and ChainRulesCore/TestUtils versions are all the same.

Hack to solve this.
Lets install the GHA that makes suggestions to fix doc-tests.
We can copy it from: https://github.com/beacon-biosignals/KeywordSearch.jl/pull/18

We kinda want to do that anyway.




If you need to get this merged can change that bit of the docs to be a  
````
```julia
````
block rather than a 
````
```@doctest
````
block.

Else can make another PR to do that stuff, and then rebase this on top of it.

Your call

**Comment by mzgubic** [comment]
> > No clue why the Documentation test fails, and I can't reproduce it locally even though the Julia version, Documenter version, and ChainRulesCore/TestUtils versions are all the same.
> 
> Hack to solve this.
> Lets install the GHA that makes suggestions to fix doc-tests.
> We can copy it from: [beacon-biosignals/KeywordSearch.jl#18](https://github.com/beacon-biosignals/KeywordSearch.jl/pull/18)
> 
> We kinda want to do that anyway.
> 
> If you need to get this merged can change that bit of the docs to be a
> 
> ```
> ```julia
> ```
> 
> block rather than a
> 
> ```
> ```@doctest
> ```
> 
> block.
> 
> Else can make another PR to do that stuff, and then rebase this on top of it.
> 
> Your call

The problem is that I can't find the place where I should make the change.😬 The doctest that is meant to be failing does not exist explicitly in the code anywhere. It looks like it was generated from `@autodocs` but the current docs don't show it either. It's a ghost.

The failure of the doctest is that the stacktrace is coming from a different line. So it looks like it is picking up some doctest that is itself using a different version of ChainRulesTestUtils (probably master?). I wonder whether we can just try merging an see if it works?

**Comment by simeonschaub** [comment]
Oh, sorry, looks like that's the doctest I introduced in #165 that's failing. I wouldn't recommend merging this before manually merging master into this branch. The failure is most likely real, because I couldn't figure out how to not make the error output I wanted to show not print the location of the testset in `test_rrule`. You could of course just correct the line number, but I agree that that's not a great solution, so I'm fine with not making that a doctest too.

**Comment by nickrobinson251** [comment]
Alternatively could you Documenters Filter functionality to ignore the line number info

**Comment by mzgubic** [comment]
> Oh, sorry, looks like that's the doctest I introduced in #165 that's failing. I wouldn't recommend merging this before manually merging master into this branch. The failure is most likely real, because I couldn't figure out how to not make the error output I wanted to show not print the location of the testset in `test_rrule`. You could of course just correct the line number, but I agree that that's not a great solution, so I'm fine with not making that a doctest too.

Oh, thanks! This was driving me crazy. I am still confused though - I thought CI runs on my branch? I haven't merged your changes. So it appears at least the Documenter action is running on what master would look like if this would have been merged? How does it deal with conflicts? 😬 How d...wh...? !??!? 😂

**Review by oxinabox — APPROVED** [review]

**Pull request merged**
